File: plugins/synthesis/tutorial/syn_1_simple.cpp
/** @file syn_1_simple.cpp

Tutorial, std monolitic synthesis. 


This tutorial uses a very simple example to illustrate the 
std monolithic controller synthesis
as originally proposed by Ramadge and Wonham.


@ingroup Tutorials 

@include syn_1_simple.cpp

*/

#include "libfaudes.h"


// we make the faudes namespace available to our program
using namespace faudes;




/////////////////
// main program
/////////////////

int main() {


  // Compose plant dynamics from two very simple machines 
  Generator tempgen, machinea, machineb;
  System cplant; 
 
  tempgen.Read("data/verysimplemachine.gen");
  tempgen.Version("1",machinea);
  tempgen.Version("2",machineb);
  Parallel(machinea,machineb,cplant);

  // Declare controllable events
  EventSet contevents;
  contevents.Insert("alpha_1");
  contevents.Insert("alpha_2");
  cplant.SetControllable(contevents);

  // Write to file
  cplant.Write("tmp_cplant12.gen");

  // Report to console
  std::cout << "################################\n";
  std::cout << "# tutorial, plant model \n";
  cplant.DWrite();
  std::cout << "################################\n";

  // Read specification 
  Generator specification;
  specification.Read("data/buffer.gen");
  InvProject(specification,cplant.Alphabet()); 
  specification.Name("simple machines specification");

  // Write to file
  specification.Write("tmp_specification12.gen");

  // Report to console
  std::cout << "################################\n";
  std::cout << "# tutorial, specification \n";
  specification.DWrite();
  std::cout << "################################\n";

  // Run synthesis algorithm
  System supervisor;
  SupCon(cplant,specification,supervisor);
  supervisor.Name("simple machines supervisor");
  supervisor.Write("tmp_supervisor12.gen");

  // Report to console
  std::cout << "################################\n";
  std::cout << "# tutorial, supervisor\n";
  supervisor.DWrite();
  std::cout << "################################\n";

  return 0;
}

File: plugins/synthesis/tutorial/syn_1_simple.py
# Synthesis demo -- two simple machines with a bfufer of size one

# import faudes module (lazy all global)
from faudes import *

# machine 1
gL1=Generator.NewFromLists(
  delta=[
    ['Idle','alpha1','Busy'],
    ['Busy','beta1','Idle']
  ],
  Q0=['Idle'],
  Qm=['Idle']
)
gL1.Name("M1")
      
# machine 2
gL2=Generator.NewFromLists(
  delta=[
    ['Idle','alpha2','Busy'],
    ['Busy','beta2','Idle']
  ],
  Q0=['Idle'],
  Qm=['Idle']
)
gL2.Name("M2")
      
# overall plant
gL=Parallel(gL1,gL2)

# controllable events
sCtrl=EventSet.NewFromList(
  ['alpha1', 'alpha2']
)

# specification aka buffer
gE=Generator.NewFromLists(
  delta=[
    ['Empty', 'beta1', 'Full'],
    ['Full', 'alpha2', 'Empty']
  ],
  Q0=['Empty'],
  Qm=['Empty']
)
gE.Name("Spec")

# lift specification to overall eventset
sAll=gL.Alphabet().Copy()
InvProject(gE,sAll,gE)

gL1.Write()
gL2.Write()
gE.Write()

# supremal closed loop
gK=SupCon(gL,sCtrl,gE)

# show result on console
gK.Write()
 
# export result as graphics
gK.GraphWrite("tmp_K.png")

# record test case
TestDump("gK",gK)

# validate test 
TestDiff()

